### PR TITLE
Config proxy [WIP]

### PR DESCRIPTION
**Pull request by OfficialNoob**
# Global config
Having a global SB.config that can be used to store, get and set variables easily.
Initiation:  `<script src="SB.js"></script>` in `<head>` already in content scripts :D
Example: `SB.config.userID = "blah";`
Debug: `SB.localconfig` (this should store all settings may be useful to enumerate in console)
Default values: `SB.defaults` from `SB.js`
Reset to default: `resetConfig()`

### CODE REVIEW

**Comment by OfficialNoob** [comment]
The code would be better in utils but its only a concept

File: popup.js [hunk 1]
```diff
@@ -1,8 +1,41 @@
+SB = {};
+
+function configProxy() {
+    chrome.storage.onChanged.addListener((changes, namespace) => {
+        for (key in changes) {
```
Comment (ajayyy) [inline popup.js]: `for (key of changes)`

Comment (OfficialNoob) [reply popup.js]: I took that from https://developer.chrome.com/extensions/storage

Comment (OfficialNoob) [reply popup.js]: Should it be changed?

Comment (ajayyy) [reply popup.js]: You are right. It seems `in` goes through a JSON object and `of` goes through an array.

**Comment by OfficialNoob** [comment]
I might try to do the Initiation in utils.js to safe time

**Comment by OfficialNoob** [comment]
Trying to get utils.js to load the sync code sooner

File: content.js [hunk 2]
```diff
@@ -72,51 +72,42 @@ var popupInitialised = false;
...
+	
 //load the legacy option to hide the notice
 var dontShowNoticeOld = false;
-chrome.storage.sync.get(["dontShowNoticeAgain"], function(result) {
-    let dontShowNoticeAgain = result.dontShowNoticeAgain;
-    if (dontShowNoticeAgain != undefined) {
-        dontShowNoticeOld = dontShowNoticeAgain;
-    }
-});
+let dontShowNoticeAgain = SB.config.dontShowNoticeAgain;
```
Comment (ajayyy) [inline content.js]: This will cause an error from multiple definitions.

Comment (OfficialNoob) [reply content.js]: Outdated

**Comment by OfficialNoob** [comment]
Added run_at for https://github.com/ajayyy/SponsorBlock/issues/221

**Comment by OfficialNoob** [comment]
@ajayyy despite the amount of code changes it seems to still work.
testing might take a while

**Comment by OfficialNoob** [comment]
SB.config.sponsorTimes.set() VS SB.config["sponsortime"+key]

**Comment by OfficialNoob** [comment]
Close was a mistake

File: SB.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            })
+        },
+        get: function(obj, prop) {
+            return SB.localconfig[prop]
+        }
+    };
+    return new Proxy({}, handler);
+}
+
+fetchConfig = _ => new Promise(function(resolve, reject) {
```
Comment (ajayyy) [inline SB.js]: Can you hange this to `(_) =>` or just `() =>` for clarity?

File: SB.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    chrome.storage.sync.get(null, function(items) {
+        SB.localconfig = items;  // Data is ready
+        resolve();
+    });
+});
+
+async function config() {
+    await fetchConfig();
+	addDefaults();
```
Comment (ajayyy) [inline SB.js]: Can you use spaces here?

File: content.js [hunk 5]
```diff
@@ -632,42 +540,30 @@ function skipToTime(v, index, sponsorTimes, openNotice) {
 
     if (openNotice) {
         //send out the message saying that a sponsor message was skipped
-        if (!dontShowNotice) {
-            let skipNotice = new SkipNotice(this, currentUUID, disableAutoSkip);
+        if (!SB.config.dontShowNotice) {
+            let skipNotice = new SkipNotice(this, currentUUID, SB.config.disableAutoSkip);
 
-            if (dontShowNoticeOld) {
+            if (!SB.config.dontShowNoticeOld) {
```
Comment (ajayyy) [inline content.js]: This old notice code can be removed now.

File: SB.js (hunk 4 above)
Comment (ajayyy) [reply SB.js]: I'll fix this after, don't worry about it.

File: SB.js (hunk 3 above)
Comment (OfficialNoob) [reply SB.js]: Yes

File: content.js (hunk 5 above)
Comment (OfficialNoob) [reply content.js]: Removed

**Comment by OfficialNoob** [comment]
> SB.config

Is this needed as It would effect performance?

**Comment by OfficialNoob** [comment]
@ajayyy this could be used for migration
```
function migrate() {
    for (key in SB.localconfig) {
        if (key.startWith("sponsortime")) {
            SB.config.set(key.substr(12), SB.config[key]);
            delete SB.config[key];
        }
    }
}
```

**Comment by ajayyy** [comment]
Either method can be used. They both should work.

**Comment by OfficialNoob** [comment]
@ajayyy Done

**Comment by OfficialNoob** [comment]
A bit to soon last time

**Comment by ajayyy** [comment]
Editing a submitted sponsor time doesn't seem to work.

**Comment by ajayyy** [comment]
Current state: editing sponsor times still doesn't work since chrome does **not** store custom classes properly. How did you get the map working?

**Comment by OfficialNoob** [comment]
Added encoding for maps however .set does not seem to cause an event

File: SB.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	return JSON.stringify(data);
+}
+
+function strParser(data) {
+	try {
+        return new Map(JSON.parse(data));
+    } catch(e) {
+        return data
+    }
+}
```
Comment (ajayyy) [inline SB.js]: Change try catch to a type check maybe? Check for type array?

Comment (OfficialNoob) [reply SB.js]: Well I think it would be a string.

Comment (OfficialNoob) [reply SB.js]: JSON.parse would return a array

**Comment by OfficialNoob** [comment]
@ajayyy I think I have added Map() support that should work.
Example: SB.config.sponsorTimes.set("x", true); (Synced)
Debug: SB.config.sponsorTimes.map (Local map)

**Comment by ajayyy** [comment]
This is probably the best way to do it

**Comment by ajayyy** [comment]
Can you please change your text editor to use 4 spaces?

**Comment by ajayyy** [comment]
I think it is ready to merge now.

Next thing to do would be to replace some of [these messages](https://github.com/ajayyy/SponsorBlock/blob/master/content.js#L124) with a listener callback from SB.config. This probably should be done in another PR however to reduce merge conflicts with the invidious PR.

**Comment by OfficialNoob** [comment]
Yeah new PR this one seems to have a lot of changes

**Pull request merged**